SimG4CMS/HcalTestBeam/plugins/HcalTB06Analysis.cc

```diff
@@ -64,44 +65,52 @@ class HcalTB06Analysis : public edm::one::EDAnalyzer<edm::one::SharedResources>
   const HcalTB06Analysis& operator=(const HcalTB06Analysis&) = delete;
 
 private:
-  edm::EDGetTokenT<edm::PCaloHitContainer> m_EcalToken;
-  edm::EDGetTokenT<edm::PCaloHitContainer> m_HcalToken;
-  edm::EDGetTokenT<edm::PCaloHitContainer> m_BeamToken;
-  bool m_ECAL;
+  const bool m_ECAL;
+  const double m_eta;
+  const double m_phi;
+  const double m_ener;
+  const std::vector<int> m_PDG;
+
+  const edm::EDGetTokenT<edm::PCaloHitContainer> m_EcalToken;
+  const edm::EDGetTokenT<edm::PCaloHitContainer> m_HcalToken;
+  const edm::EDGetTokenT<edm::PCaloHitContainer> m_BeamToken;
+
+  const edm::ParameterSet m_ptb;
+  const double m_timeLimit;
+  const double m_widthEcal;
+  const double m_widthHcal;
+  const double m_factEcal;
+  const double m_factHcal;
+  const double m_eMIP;
 
   int count;
   int m_idxetaEcal;
   int m_idxphiEcal;
   int m_idxetaHcal;
   int m_idxphiHcal;
 
-  double m_eta;
-  double m_phi;
-  double m_ener;
-  double m_timeLimit;
-  double m_widthEcal;
-  double m_widthHcal;
-  double m_factEcal;
-  double m_factHcal;
-  std::vector<int> m_PDG;
-
-  HcalTB06Histo* m_histo;
+  std::unique_ptr<HcalTB06Histo> m_histo;
 };
 
-HcalTB06Analysis::HcalTB06Analysis(const edm::ParameterSet& p) : count(0) {
+HcalTB06Analysis::HcalTB06Analysis(const edm::ParameterSet& p)
+    : m_ECAL(p.getParameter<bool>("ECAL")),
+      m_eta(p.getParameter<double>("MinEta")),
+      m_phi(p.getParameter<double>("MinPhi")),
+      m_ener(p.getParameter<double>("MinE")),
+      m_PDG(p.getParameter<std::vector<int> >("PartID")),
+      m_EcalToken(consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits", "EcalHitsEB"))),
+      m_HcalToken(consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits", "HcalHits"))),
+      m_BeamToken(consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits", "HcalTB06BeamHits"))),
+      m_ptb(p.getParameter<edm::ParameterSet>("TestBeamAnalysis")),
+      m_timeLimit(m_ptb.getParameter<double>("TimeLimit")),
+      m_widthEcal(m_ptb.getParameter<double>("EcalWidth")),
+      m_widthHcal(m_ptb.getParameter<double>("HcalWidth")),
+      m_factEcal(m_ptb.getParameter<double>("EcalFactor")),
+      m_factHcal(m_ptb.getParameter<double>("HcalFactor")),
+      m_eMIP(m_ptb.getParameter<double>("MIP")),
+      count(0) {
   usesResource("TFileService");
 
-  m_ECAL = p.getParameter<bool>("ECAL");
```

Comment:
@bsunanda , this "if" before consumes is needed for calibration, without this would be not possible.

Reply:
Previous code was good enough.